homeassistant/components/withings/__init__.py

```diff
@@ -62,52 +80,126 @@ async def async_setup(hass: HomeAssistantType, config: ConfigType) -> bool:
-    if "auth_implementation" not in entry.data:
-        _LOGGER.debug("Upgrading existing config entry")
-        data = entry.data
-        creds = data.get(CREDENTIALS, {})
-        hass.config_entries.async_update_entry(
-            entry,
-            data={
-                "auth_implementation": DOMAIN,
-                "implementation": DOMAIN,
-                "profile": data.get("profile"),
-                "token": {
-                    "access_token": creds.get("access_token"),
-                    "refresh_token": creds.get("refresh_token"),
-                    "expires_at": int(creds.get("token_expiry")),
-                    "type": creds.get("token_type"),
-                    "userid": creds.get("userid") or creds.get("user_id"),
-                },
+    config_updates = {}
+
+    # Add a unique id if it's an older config entry.
+    if entry.unique_id != entry.data["token"]["userid"]:
+        config_updates["unique_id"] = entry.data["token"]["userid"]
+
+    # Add the webhook configuration.
+    if CONF_WEBHOOK_ID not in entry.data:
+        webhook_id = webhook.async_generate_id()
+        config_updates["data"] = {
+            **entry.data,
+            **{
+                const.CONF_USE_WEBHOOK: hass.data[DOMAIN][const.CONFIG][
+                    const.CONF_USE_WEBHOOK
+                ],
+                CONF_WEBHOOK_ID: webhook_id,
+                const.CONF_WEBHOOK_URL: entry.data.get(
+                    const.CONF_WEBHOOK_URL,
+                    webhook.async_generate_url(hass, webhook_id),
+                ),
             },
-        )
+        }
 
-    implementation = await config_entry_oauth2_flow.async_get_config_entry_implementation(
-        hass, entry
+    if config_updates:
+        hass.config_entries.async_update_entry(entry, **config_updates)
+
+    data_manager = await async_get_data_manager(hass, entry)
+
+    _LOGGER.debug("Confirming %s is authenticated to withings", data_manager.profile)
+    await data_manager.poll_data_update_coordinator.async_refresh()
+    if not data_manager.poll_data_update_coordinator.last_update_success:
+        raise ConfigEntryNotReady()
+
+    webhook.async_register(
+        hass,
+        const.DOMAIN,
+        "Withings notify",
+        data_manager.webhook_config.id,
+        async_webhook_handler,
     )
 
-    data_manager = get_data_manager(hass, entry, implementation)
+    # Perform first webhook subscription check.
+    if data_manager.webhook_config.enabled:
+        data_manager.async_start_polling_webhook_subscriptions()
 
-    _LOGGER.debug("Confirming we're authenticated")
-    try:
-        await data_manager.check_authenticated()
-    except NotAuthenticatedError:
-        _LOGGER.error(
-            "Withings auth tokens exired for profile %s, remove and re-add the integration",
-            data_manager.profile,
-        )
-        return False
+        def call_later_callback(now) -> None:
```

Comment:
This needs to be decorated with `@callback` imported from core.py. Otherwise it will run in sync context. Please name it `async_*`.